File: mvc-arch/src/main/java/org/example/mvc/DispatcherServlet.java
package org.example.mvc;

import org.example.mvc.controller.Controller;
import org.example.mvc.view.JspViewResolver;
import org.example.mvc.view.View;
import org.example.mvc.view.ViewResolver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

@WebServlet("/")
public class DispatcherServlet extends HttpServlet {
    private static final Logger log = LoggerFactory.getLogger(DispatcherServlet.class);
    private RequestMappingHandlerMapping handlerMapping;
    private List<ViewResolver> viewResolvers;
    private List<HandlerAdapter> handlerAdapters;

    @Override
    public void init() throws ServletException {
        handlerMapping = new RequestMappingHandlerMapping();
        handlerMapping.init();

        viewResolvers = Collections.singletonList(new JspViewResolver());
        handlerAdapters = List.of(new SimpleControllerHandlerAdapter());
    }

    @Override
    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        log.info("[DispatcherServlet] service started");

        try {
            RequestMethod method = RequestMethod.valueOf(req.getMethod());
            String path = req.getRequestURI();

            Controller handler = handlerMapping.findHandler(new HandlerKey(method, path));

            HandlerAdapter handlerAdapter = handlerAdapters.stream()
                    .filter(ha -> ha.supports(handler))
                    .findFirst()
                    .orElseThrow(() -> new ServletException("No Adapter"));

            ModelAndView modelAndView = handlerAdapter.handle(req, resp, handler);

            for (ViewResolver viewResolver : viewResolvers) {
                View view = viewResolver.resolveView(modelAndView.getViewName());
                view.render(modelAndView.getModel(), req, resp);
            }

        } catch (Exception e) {
            log.error("Exception occurred : [{}]", e.getMessage());
        }
    }
}

File: refactor-to-kotlin/src/main/kotlin/org/example/mvc/DispatcherServlet.kt
package org.example.mvc

import org.example.mvc.adapter.HandleAdapter
import org.example.mvc.adapter.SimpAdapter
import org.example.mvc.handlerMapping.HandlerKey
import org.example.mvc.handlerMapping.HandlerMapping
import org.example.mvc.handlerMapping.RequestMappingHandlerMapping
import org.example.mvc.view.View
import org.example.mvc.view.resolver.JspViewResolver
import org.example.mvc.view.resolver.ViewResolver
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import javax.servlet.annotation.WebServlet
import javax.servlet.http.HttpServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@WebServlet("/")
class DispatcherServlet : HttpServlet() {
    private lateinit var log : Logger
    private lateinit var handlerMappings : List<HandlerMapping>
    private lateinit var handleAdapters : List<HandleAdapter>
    private lateinit var viewResolvers : List<ViewResolver>

    override fun init() {
        log = LoggerFactory.getLogger(DispatcherServlet::class.java)
        log.debug("init()")

        val requestMappingHandlerMapping = RequestMappingHandlerMapping()
        requestMappingHandlerMapping.init()


        handlerMappings = listOf(requestMappingHandlerMapping)
        handleAdapters = listOf(SimpAdapter())
        viewResolvers = listOf(JspViewResolver())
    }

    override fun service(req: HttpServletRequest, resp: HttpServletResponse) {
        log.debug("service()")

        val requestURI = req.requestURI
        val method = RequestMethod.valueOf(req.method)
        val handlerKey = HandlerKey(requestURI, method)

        val handler = handlerMappings
            .first { it.findHandler(handlerKey) != null }
            .findHandler(handlerKey)

        val adapter = handleAdapters.first { it.supports(handler) }
        val modelAndView = adapter.handle(req, resp, handler)

        for (viewResolver in viewResolvers) {
            val view : View = viewResolver.resolveView(modelAndView.getViewName()!!)
            view.render(modelAndView.model, req, resp)
        }
    }
}